### ISSUE TITLE
Mapped port not reachable on all ethernet interfaces / IP:Port port mapping not saved

### ISSUE THREAD

**Issue by janeczku** [rancherio/rancher#508]
**Steps to reproduce**
Prerequisite: Two ethernet interfaces (eth0: public network, eth1: private network)
1. Click on "Add Container"
2. Next to "Select image", select "Docker" and enter "tutum/hello-world"
3. Next to "Port Map", enter **Host_eth0_IP:80** in the public and **80** in the private field
4. Click "Create Container"
5. When container has started, try open the website at **http://Host_eth0_IP:80** ( -> "Connection refused")
6. Try open the website at **http://Host_eth1_IP:80** ( -> "Connection refused")
7. Open the menu for Host->Container->hello-world and select "Edit"
8. Check values of "Port Bindings" (-> They are empty!)
9. Enter **80** in both fields and click on "Save"
10. Try open the website at **http://Host_eth0_IP:80** ( -> "Connection refused")
11. Try open the website at **http://Host_eth1_IP:80** ( -> "Connection accepted")

**Expected result**
1. Entering the public port in the form of HOST_IP:PORT is working
2. The container is reachable under eth0 public IP

**Actual result**
1. The port mapping is not saved
2. The container is only reachable under eth1 private IP

**Additional info**
```
core@coreos-par-35 / $ cat /etc/os-release
NAME=CoreOS
VERSION=607.0.0
```

```
core@coreos-par-35 / $ docker ps
CONTAINER ID        IMAGE                           COMMAND                CREATED             STATUS              PORTS                                          NAMES
22d2d748d4cd        tutum/hello-world:latest        "/run.sh"              8 minutes ago       Up 8 minutes        80/tcp                                         db7d65a6-5966-46b7-b8aa-040ca4c7890b   
d907525fe968        rancher/agent-instance:v0.3.0   "/etc/init.d/agent-i   9 minutes ago       Up 9 minutes        0.0.0.0:500->500/udp, 0.0.0.0:4500->4500/udp   0a1332fe-9067-445c-939c-8ac6c3e77731   
ec78c1f523ea        rancher/agent:v0.5.2            "/run.sh run"          12 minutes ago      Up 12 minutes   
```

```
core@coreos-par-35 /etc/systemd/system $ ifconfig
docker0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.42.1  netmask 255.255.0.0  broadcast 0.0.0.0
        inet6 fe80::5484:7aff:fefe:9799  prefixlen 64  scopeid 0x20<link>
        ether 56:84:7a:fe:97:99  txqueuelen 0  (Ethernet)
        RX packets 5708  bytes 1137091 (1.0 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 10201  bytes 5052672 (4.8 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

dummy0: flags=195<UP,BROADCAST,RUNNING,NOARP>  mtu 1500
        inet6 fe80::c00a:51ff:fe36:b0a4  prefixlen 64  scopeid 0x20<link>
        ether c2:0a:51:36:b0:a4  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 381  bytes 123060 (120.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 108.61.xxx.xxx  netmask 255.255.254.0  broadcast 108.61.xxx.xxx
        inet6 2001:19f0:6800:85a6:5400:ff:fe07:caf1  prefixlen 64  scopeid 0x0<global>
        inet6 2001:19f0:6800:85a6::64  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::5400:ff:fe07:caf1  prefixlen 64  scopeid 0x20<link>
        ether 56:00:00:07:ca:f1  txqueuelen 1000  (Ethernet)
        RX packets 89627  bytes 572913267 (546.3 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 82693  bytes 23729252 (22.6 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.99.0.10  netmask 255.255.0.0  broadcast 0.0.0.0
        inet6 fe80::5054:ff:fe90:40b5  prefixlen 64  scopeid 0x20<link>
        ether 52:54:00:90:40:b5  txqueuelen 1000  (Ethernet)
        RX packets 802  bytes 241488 (235.8 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 416  bytes 124063 (121.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 0  (Local Loopback)
        RX packets 4549  bytes 12122368 (11.5 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4549  bytes 12122368 (11.5 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

veth771eac9: flags=67<UP,BROADCAST,RUNNING>  mtu 1500
        inet6 fe80::2c33:85ff:fe88:bb7  prefixlen 64  scopeid 0x20<link>
        ether 2e:33:85:88:0b:b7  txqueuelen 0  (Ethernet)
        RX packets 8  bytes 648 (648.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 8  bytes 648 (648.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

```
core@coreos-par-35 /etc/systemd/system $ sudo iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain DOCKER (1 references)
target     prot opt source               destination  
```

```
core@coreos-par-35 /etc/systemd/system $ sudo iptables -t nat -L   
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
CATTLE_PREROUTING  all  --  anywhere             anywhere            
DOCKER     all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere            !loopback/8           ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
CATTLE_POSTROUTING  all  --  anywhere             anywhere            
MASQUERADE  all  --  172.17.0.0/16        anywhere            

Chain CATTLE_POSTROUTING (1 references)
target     prot opt source               destination         
MASQUERADE  tcp  --  10.42.0.0/16        !10.42.0.0/16         masq ports: 1024-65535
MASQUERADE  udp  --  10.42.0.0/16        !10.42.0.0/16         masq ports: 1024-65535
MASQUERADE  all  --  10.42.0.0/16        !10.42.0.0/16        
MASQUERADE  tcp  --  172.17.0.0/16        anywhere             masq ports: 1024-65535
MASQUERADE  udp  --  172.17.0.0/16        anywhere             masq ports: 1024-65535

Chain CATTLE_PREROUTING (1 references)
target     prot opt source               destination         
DNAT       udp  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL udp dpt:ipsec-nat-t to:10.42.231.255:4500
DNAT       udp  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL udp dpt:isakmp to:10.42.231.255:500
DNAT       all  --  10.42.0.0/16         169.254.169.254      to:10.42.231.255

Chain DOCKER (2 references)
target     prot opt source               destination 
```
*docker logs rancher/agent:v0.5.2*
```
E0412 11:14:50.781351     717 error.go:26] %q
%s
 unable to unmarshal container info for "docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a" (failed to get container "/docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a" with error: unknown container "/docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a"
): invalid character 'i' in literal false (expecting 'l')
2015/04/12 11:14:50 http: response.WriteHeader on hijacked connection
2015/04/12 11:14:50 http: response.Write on hijacked connection
I0412 11:14:50.781452     717 access.go:43] 46.128.xxx.xxx HTTP/1.1 GET: /v1/stats/db7d65a6-5966-46b7-b8aa-040ca4c7890b?token=eyJhbGciOiJSUzI1NiJ9.eyJwb3J0Ijo5MzQ1LCJleHAiOjE0Mjg4Mzc1OTAsInN1YiI6ImNhdHRsZSIsImlzcyI6Imh0dHA6XC9cL2NhdHRsZS5pbyIsImhvc3RVdWlkIjoiNGJkNGVhY2ItOGFiNC00NTEyLTg5MjctZDZjOThkZTVmYjYyIiwiaWF0IjoxNDI4ODM3MjkwLCJpcEFkZHJlc3MiOiIxMDguNjEuMjA5LjkyIiwia2lkIjoiZGVmYXVsdCJ9.CxOly9nJcbFMyfgYlhVC4EEviGYRENEb2z7XWwZFyDY1L79AuUA_RTESm2rHNgcNCEP_SBvxzeI-KjOI2HzZtZ7XAfSS_Rgx6xbXa_wYxCQ0kIWmuygn_VE-zh1hnxrNjjEAy4KmXOEioti1Gbp-KR85EKEF3heC02axyZho4BnnZUIwklqiOtKqbPjSC2dKDHScM32e4OglANN3KNbcY0MD8SgF_Li755HHDz828A0Rv0O3ysMRpWKt8KieHEOSP9JbWxtMrDTwU4Tq5zqdpX1NOjBZOj5o7vTNjhYqcjW40YyNTJycFjQBTOKDWPZ5KodE2RQeepelza4Eioapgw, host: 108.61.XXX.XXX:9345 (load time: 0.00266 seconds)
2015/04/12 11:14:52 "unable to unmarshal container info for \"docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a\" (failed to get container \"/docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a\" with error: unknown container \"/docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a\"\n): invalid character 'i' in literal false (expecting 'l')"
goroutine 33 [running]:
github.com/rancherio/host-api/app/common.CheckError(0x7f54a75f9940, 0xc2080acb30, 0x2)
	/source/gopath/src/github.com/rancherio/host-api/app/common/error.go:17 +0x8a
github.com/rancherio/host-api/app/common.ErrorHandler.ServeHTTP(0x8947b0, 0x7f54a7600ed8, 0xc2080565a0, 0xc20800a750)
	/source/gopath/src/github.com/rancherio/host-api/app/common/error_handler.go:13 +0x8a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc20802e410, 0x7f54a7600ed8, 0xc2080565a0, 0xc20800a750)
	/source/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x297
github.com/rancherio/host-api/auth.func·002(0x7f54a7600ed8, 0xc2080565a0, 0xc20800a750)
	/source/gopath/src/github.com/rancherio/host-api/auth/auth.go:55 +0xfc
net/http.HandlerFunc.ServeHTTP(0xc20802aef0, 0x7f54a7600ed8, 0xc2080565a0, 0xc20800a750)
	/usr/src/go/src/net/http/server.go:1265 +0x41
net/http.(*ServeMux).ServeHTTP(0xc20803c780, 0x7f54a7600ed8, 0xc2080565a0, 0xc20800a750)
	/usr/src/go/src/net/http/server.go:1541 +0x17d
net/http.serverHandler.ServeHTTP(0xc2080548a0, 0x7f54a7600ed8, 0xc2080565a0, 0xc20800a750)
	/usr/src/go/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc208056000)
	/usr/src/go/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
	/usr/src/go/src/net/http/server.go:1751 +0x35e

E0412 11:14:52.199444     717 error.go:26] %q
%s
 unable to unmarshal container info for "docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a" (failed to get container "/docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a" with error: unknown container "/docker/22d2d748d4cd9a8189f3a783b749b13078c07b22d2266c5cd590ea735d608d9a"
): invalid character 'i' in literal false (expecting 'l')
2015/04/12 11:14:52 http: response.WriteHeader on hijacked connection
2015/04/12 11:14:52 http: response.Write on hijacked connection
I0412 11:14:52.199515     717 access.go:43] 46.128.xxx.xxx HTTP/1.1 GET: /v1/stats/db7d65a6-5966-46b7-b8aa-040ca4c7890b?token=eyJhbGciOiJSUzI1NiJ9.eyJwb3J0Ijo5MzQ1LCJleHAiOjE0Mjg4Mzc1OTAsInN1YiI6ImNhdHRsZSIsImlzcyI6Imh0dHA6XC9cL2NhdHRsZS5pbyIsImhvc3RVdWlkIjoiNGJkNGVhY2ItOGFiNC00NTEyLTg5MjctZDZjOThkZTVmYjYyIiwiaWF0IjoxNDI4ODM3MjkwLCJpcEFkZHJlc3MiOiIxMDguNjEuMjA5LjkyIiwia2lkIjoiZGVmYXVsdCJ9.CxOly9nJcbFMyfgYlhVC4EEviGYRENEb2z7XWwZFyDY1L79AuUA_RTESm2rHNgcNCEP_SBvxzeI-KjOI2HzZtZ7XAfSS_Rgx6xbXa_wYxCQ0kIWmuygn_VE-zh1hnxrNjjEAy4KmXOEioti1Gbp-KR85EKEF3heC02axyZho4BnnZUIwklqiOtKqbPjSC2dKDHScM32e4OglANN3KNbcY0MD8SgF_Li755HHDz828A0Rv0O3ysMRpWKt8KieHEOSP9JbWxtMrDTwU4Tq5zqdpX1NOjBZOj5o7vTNjhYqcjW40YyNTJycFjQBTOKDWPZ5KodE2RQeepelza4Eioapgw, host: 108.61.XXX.XXX:9345 (load time: 0.00245 seconds)
```
*docker logs rancher/agent-instance:v0.3.0*
http://pastebin.com/B96jGDEG

**Comment by ibuildthecloud** [comment]
After step 7 does it work or still fail?

**Comment by janeczku** [comment]
it still fails.

**Comment by ibuildthecloud** [comment]
Thanks, I'll investigate.  Based on the iptables rules it is clear the port mapping is not being configured.

**Comment by janeczku** [comment]
I investigated some more. This problem is obviously related to the host having multiple physical interfaces.

**Comment by janeczku** [comment]
I have amended the information in the OP.

**Comment by ibuildthecloud** [comment]
Right now publishing a port to a specific IP is not supported.  It should be and its just a bug its not.

Does the functionality work if you do the equivalent of "-p 80:80"

**Comment by janeczku** [comment]
Whoa, suddenly it started working when entering 80:80 for port mapping. Very strange, since thats the way i did it when launching the container in the first place. I will try to reproduce the problem on a new host and write the result here.

**Comment by janeczku** [comment]
Okay so i cannot reproduce the original problem. You might probably close this issue.

**Comment by osobh** [comment]
I'm having the same issue when creating containers and mapping the IP. 

With a container created in Rancher I can see from the cli that the ip mappings look like:
 1337/tcp, 7474/tcp

While other containers such as the agent show their mappings as:

0.0.0.0:500->500/udp, 0.0.0.0:4500->4500/udp

so it seems like when the container is created it's not mapping the needed ports to the containers public interface. The mapping of the container I want to start should look like after rancher creates the container:
0.0.0.0:7474->7474/tdp, 0.0.0.0:1337->1337/tcp

Hope this helps

**Comment by deniseschannon** [comment]
Duplicate of https://github.com/rancherio/rancher/issues/53.
